ForumProject/users/views.py

```diff
@@ -60,6 +66,76 @@ def post(self, request, *args, **kwargs):
         return response
 
 
+class RoleSelectionView(APIView):
+    permission_classes = [IsAuthenticated]
+
+    def post(self, request):
+        serializer = RoleSerializer(data=request.data)
+        if serializer.is_valid():
+            user = request.user
+            if UserRoleCompany.objects.filter(user=user).exists():
+                user_role = UserRoleCompany.objects.get(user=user)
+                user_role.role = serializer.validated_data['role']
+                user_role.company_id = 0
+                user_role.save()
+            else:
+                UserRoleCompany.objects.create(user=user, role=serializer.validated_data['role'])
+            return Response({'success': 'Role has been successfully updated.'}, status=status.HTTP_200_OK)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+
+class CompanySelectionView(APIView):
+    permission_classes = [IsRole]
+
+    def post(self, request):
+        serializer = CompanySerializer(data=request.data)
+        if serializer.is_valid():
+            user = request.user
+            company = serializer.validated_data['company_id']
+
+            user_role_company = UserRoleCompany.objects.get(user=user)
+
+            if user_role_company.role == 'investor':
+                user_investors = UserInvestor.objects.filter(customuser=self.request.user).values('investor')
+                if {'investor': company} in user_investors:
+                    user_role_company.company_id = company
+                    user_role_company.save()
+                    return Response({'success': 'The investor company was successfully selected'}, status=status.HTTP_200_OK)
+
+            elif user_role_company.role == 'startup':
+                user_startups = UserStartup.objects.filter(customuser=self.request.user).values('startup')
+                if {'startup': company} in user_startups:
+                    user_role_company.company_id = company
+                    user_role_company.save()
+                    return Response({'success': 'The startup company was successfully selected'}, status=status.HTTP_200_OK)
+            return Response({'error': 'You must select only your company'}, status=status.HTTP_400_BAD_REQUEST)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```

Comment:
1. Similar to 'RoleSelectionView', use more efficient queries. For example, instead of fetching all user's companies and then checking if a company is in that list, consider querying directly for the specific company
2. You should handle cases where `UserRoleCompany.objects.get(user=user)` might not find an entry, which will raise a `DoesNotExist` exception. Use `get_object_or_404` or handle this case explicitly.

Reply:
Thanks,  very useful advice.